### ISSUE TITLE
`HttpError` or `HttpError: Bad Credentials`

### ISSUE THREAD

**Issue by ianwremmel** [octokit/octokit.js#1831]
**Checklist**

- [X] I checked [all open bugs](https://github.com/octokit/rest.js/issues?q=is%3Aissue+label%3Abug) and none of them matched my problem.

**Environment**

- [X] https://api.github.com
- [ ] GitHub Enterprise Server: 
- [ ] GitHub Enterprise Cloud
- [X] Node.js
- [ ] Browsers: 
- [ ] Others: 

**Versions**

```
├─┬ @octokit/app@4.2.1
│ ├─┬ @octokit/request@5.3.1
│ │ ├─┬ @octokit/endpoint@5.5.1
│ │ ├─┬ @octokit/types@2.0.2
├─┬ @octokit/plugin-retry@2.2.0
├─┬ @octokit/plugin-throttling@2.7.1
├─┬ @octokit/rest@16.43.1
│ ├─┬ @octokit/auth-token@2.4.0
│ ├─┬ @octokit/plugin-paginate-rest@1.1.2
│ ├── @octokit/plugin-request-log@1.0.0
│ ├─┬ @octokit/plugin-rest-endpoint-methods@2.4.0
│ ├─┬ @octokit/request-error@1.0.4
│ ├── octokit-pagination-methods@1.1.0
├─┬ @octokit/webhooks@7.6.5
```

> I'm still on rest@16 because the change in type definitions has been an insurmountable hurdle.

**What happened?**

I'm running a GitHub app that, as recommnded by the marketplace docs, syncs with GitHub every day. About two weeks ago, I started getting `HttpError: Bad Credentials` pretty regularly. Mostly, it happened during syncs but it also occasionally happened during regular API calls. Retries worked around the issue. 

A few days ago, that sync job started failing far more agresssively - I now just get `HttpError`, retries don't fix it, and I'm pretty sure none of the syncs succeed.

I recently added `bottleneck` to see if I was hitting some kind of credential limit, but I'm fairly confident I'm not creating more credentials than I need to. 

**Minimal test case to reproduce the problem**

I'm not really sure how to reproduce it yet. This stack trace isn't particularly helpful:

![Screen Shot 2020-08-07 at 9 05 45 AM](https://user-images.githubusercontent.com/1182361/89665254-3c302d00-d88d-11ea-8554-102a52302ba7.png)

**What did you expect to happen?**

I expected I wouldn't get an `HttpError`

**What the problem might be**

The only thing I can think of is that I'm authorizing a few hundred installations in a short period and then making a few calls using each set of credentials, but I feel like I have a far lower magnitude than many other apps, so that seems unlikely.

**Comment by gr2m** [comment]
You use an old version of `@octokit/rest` with a built-in authentication for GitHub apps which is no longer supported. I think that the problem is that this version is hitting a deprecated API endpoint which is being browned out right now, meaning that GitHub turns it off more and more frequently. See https://developer.github.com/changes/2020-04-15-replacing-create-installation-access-token-endpoint/. You should have received lots of emails about this with the subject "[GitHub API] Deprecation notice for Installation Access Token creation API"

Can you please upgrade to the latest `@octokit/rest` and use https://github.com/octokit/auth-app.js/?

**Comment by ianwremmel** [comment]
> Can you please upgrade to the latest @octokit/rest and use https://github.com/octokit/auth-app.js/?

not really. the new typescript definitions have made that migration all but impossible. 

> You should have received lots of emails about this with the subject "[GitHub API] Deprecation notice for Installation Access Token creation API"

Hmm, I got a bunch of emails about a deprecated event going away, but I don't seem to have gotten that one. That blog post does seem to explain the exact problem though.

**Comment by gr2m** [comment]
> the new typescript definitions have made that migration all but impossible

Can you clarify?

**Comment by ianwremmel** [comment]
pretty much all the concerns I and others have raised in issues about the new types. I'm using octokit's types in a lot of places throughout my app, so it's very big refactor to get access to the new type definitions. I've attempted it several times and given up after a few days each time.

**Comment by gr2m** [comment]
I won't be able to help with the issue because you use very outdated versions. The real issue is the export of the TypeScript types for parameters and responses. I'll get to work on that eventually, but it's not my priority right now, sorry.

**Comment by ianwremmel** [comment]
Ok, I've managed to upgrade and 18, but I'm still working on switching to `createAppAuth`. Does it replace `@octokit/app`? It looks like their Readmes each indicate they solve the same problem. Why doesn't `@octokit/app` build `createAppAuth` in?

**Comment by gr2m** [comment]
Great!

I will reuse `@octokit/app` for an GitHub App library in future, please use `@octokit/auth-app`. 

Starting with v17, `@octokit/rest` is built on https://github.com/octokit/core.js, which accepts two authentication related parameters

1. `authStrategy`
2. `auth`

You set `authStrategy` to `createAppAuth`, and `auth` to `{ id, privateKey, installationId}`. The correct means of authentication (JWT or access token) is used automatically based on the request URL.

**Comment by ianwremmel** [comment]
Hmm, as far as I can tell, I'm successfully setup auth-app according to its README, but I still get the following error in Sentry when the sync job runs during the brownout window.

```
HttpError: 
    at /var/app/current/node_modules/@octokit/request/dist-src/fetch-wrapper.js:49:31
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

**Comment by gr2m** [comment]
That error message is not very helpful 🤔 is there any way you can find out more about the error?

**Comment by ianwremmel** [comment]
haha, yea, I was hoping you had some ideas there :) 

That's what Sentry is grabbing. I can try to hook into the Sentry error reporter and capture more fields, but I'm not sure what there is to capture. Do you know what interesting fields get assigned to http errors? I seem to remember there's an `Object.assign` in there somewhere. 

Is octokit just using native Promises or is there a polyfill that's causing the lost of the stack trace? I can try to enable async stack traces.

**Comment by gr2m** [comment]
at least it should have gotten `error.message`, it's odd that it seems to be empty?

If the error originated from an request, then you get an [`@octokit/request-error`](https://github.com/octokit/request-error.js/)  with these extra properties

```js
error.status; // e.g. 500
error.headers; // e.g. { 'x-github-request-id': '1:2:3:4' }
error.request.method; // e.g. POST
error.request.url; // e.g. https://api.github.com/foo
error.request.body; // e.g. { bar: 'baz' }
error.request.headers; // e.g. { authorization: 'token [REDACTED]' }
```

**Comment by ianwremmel** [comment]
> at least it should have gotten error.message, it's odd that it seems to be empty?

yea, that's what I was thinking. That happened sometime last week; I had been getting the "Bad Credentials" error, but it changed to just HttpError.

**Comment by ianwremmel** [comment]
I think I found the auth problem. Turns out I missed one of the spots where I was still using the old system. Will hopefully get it merged before the next sync runs, but if not, we should at least get more details about the HttpError.

**Comment by ianwremmel** [comment]
ok, the last auth fix isn't deployed yet, so I got more info on the HttpError:

headers

```
{
    access-control-allow-origin: *, 
    access-control-expose-headers: [Filtered], 
    connection: close, 
    content-length: 0, 
    content-security-policy: default-src 'none', 
    content-type: application/json; charset=utf-8, 
    date: Wed, 12 Aug 2020 03:08:15 GMT, 
    referrer-policy: origin-when-cross-origin, strict-origin-when-cross-origin, 
    server: GitHub.com, 
    status: 500 Internal Server Error, 
    strict-transport-security: max-age=31536000; includeSubdomains; preload, 
    vary: Accept-Encoding, Accept, X-Requested-With, 
    x-content-type-options: nosniff, 
    x-frame-options: deny, 
    x-github-media-type: github.v3; format=json, 
    x-github-request-id: E74E:19AE:2F09DD:819993:5F335D1F, 
    x-ratelimit-limit: 5000, 
    x-ratelimit-remaining: 4994, 
    x-ratelimit-reset: 1597205279, 
    x-xss-protection: 1; mode=block
}
```

request	

```
{
    headers: {
        accept: application/vnd.github.v3+json, 
        authorization: [Filtered], 
        user-agent: octokit-rest.js/18.0.3 octokit-core.js/3.1.2 Node.js/12.18.3 (linux; x64)
    }, 
    method: GET, 
    request: {
        retries: 3, 
        retryAfter: 16, 
        retryCount: 3
    }, 
    url: https://api.github.com/marketplace_listing/plans/2485/accounts?page=2
}
```

status

```
500
```


The stack trace is still unhelpful though.

**Comment by gr2m** [comment]
A 500 error is not easy to produce, it might be a problem with the API itself. I’ll give it a try tomorrow

**Comment by ianwremmel** [comment]
Thanks. It does seem to be happening pretty consistently every night when the sync job runs. It's interesting that the url is calling for page 2, implying the first page loaded, but the second fails.

**Comment by ianwremmel** [comment]
Found the issue, sort of. It looks like there's a bad piece of data in there. Through a bunch of manual testing, I found that if I use a page size of 1, I only get the error on what is currently page 41. My guess is, that's the user that was also causing the unique id constraint in my database for two users having the same user name. I'll open a support ticket directly with GitHub since this seems to be a server-side problem.
